Library/Homebrew/livecheck/strategy.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      module_function
+
+      def strategies
```

Comment:
This makes me think having a static list of strategies defined somewhere (which can be added to from taps, if needed) is a better fit than doing introspection here. We can add a unit test to ensure that the list doesn't fall out of date.

Reply:
One thing that's important is we only want a tap's strategies (or strategy overrides) to apply to formulae in that tap. We currently load tap strategies when there are formulae present from non-core taps but these strategies/overrides apply to all formulae in the run, which we don't want.

For example, I have a local test tap where I've implemented a `GithubLatest` strategy (which checks the "latest" release on GitHub). If I set the priority higher than others (e.g., 9 or 10) and run a check that includes a formula in my tap as well as homebrew-core formulae, then the strategy will be applied to all the formulae that match in a run (not just the tap formula).

Another example would be if a tap had a strategy that simply modified some aspects of a built-in strategy:

```ruby
module Homebrew
  module Livecheck
    module Strategy
      module Git
        PRIORITY = 4 # Instead of 8 in the built-in strategy
      end
    end
  end
end
```

This modification would apply in the same manner, affecting all the formulae in a run if a formula in the tap is included.

This is also a problem if two taps implement a strategy with the same name, where one would override the other (or we end up with a mix of the two).

---

The basic idea is that we need to differentiate between built-in strategies and tap strategies and only apply tap strategies (including any modifications to existing strategies) to formulae from that tap. I'm not yet sure how we can accomplish this from a technical standpoint but I imagine it will affect how we proceed with this.

Reply:
> This is also a problem if two taps implement a strategy with the same name

Maybe the name can be scoped under the tap name in that case?

Also, it seems like strategies should really be classes instead of modules so you can subclass them.

Reply:
> Maybe the name can be scoped under the tap name in that case?
> 
> Also, it seems like strategies should really be classes instead of modules so you can subclass them.

Agreed 👍🏻

Reply:
> Maybe the name can be scoped under the tap name in that case?

Yes, that's what I'm thinking. There may be some challenges around implementing this but we can flesh it out in a separate issue or PR after GSoC.

> Also, it seems like strategies should really be classes instead of modules so you can subclass them.

I agree. I originally had them as classes when I established the structure of the strategies in the homebrew-livecheck repository, for this same reason. The strategies became modules related to [a requested change in #8180](https://github.com/Homebrew/brew/pull/8180#discussion_r464260242) (when the strategy changes were part of that PR), seemingly to take advantage of `module_function`.

Do we want to work on converting the strategies back to classes before merging this PR or should we simply move this forward and handle it later (due to time constraints)?

Reply:
> should we simply move this forward and handle it later (due to time constraints)?

The latter. The only thing I see blocking is @reitermarkus's documentation request below.